test/ERC4626Adapter.test.ts

```diff
@@ -223,162 +226,283 @@ describe('ERC4626 Adapter', () => {
+      })
 
-          before('deposit 30 assets for userB', async () => {
-            await token.connect(userB).approve(erc4626Adapter.address, amount)
-            await erc4626Adapter.connect(userB).deposit(amount, userB.address)
+      context('when assets are drained', async () => {
+        const itWorksAsExpected = (drainAmount: BigNumber, maxWithdrawAmount: BigNumber, fee: BigNumberish) => {
+          beforeEach('drain assets', async () => {
+            await token.burn(erc4626Mock.address, drainAmount)
           })
 
-          totalAssets = totalAssets.add(amount) // 330
-          userBAssets = amount // 30
-          userBShares = amount.mul(fp(1)).div(shareValue) // 10.714
-          totalShares = totalShares.add(userBShares) // 117.854
-          previousTotalAssets = totalAssets // 330
-
-          checkStatus({
-            totalAssets,
-            shareValue,
-            userAShares,
-            userAAssets,
-            userBShares,
-            userBAssets,
-            userCShares,
-            userCAssets,
-            totalShares,
-            previousTotalAssets,
-          })
+          context('when userA withdraws the assets left', async () => {
+            const amount = maxWithdrawAmount
 
-          context('when assets duplicate', async () => {
-            const amount = fp(330)
+            it('lets users withdraw all their assets left', async function () {
+              const previousBalance = await token.balanceOf(userA.address)
 
-            before('duplicate assets', async () => {
-              await token.mint(erc4626Mock.address, amount)
+              await erc4626Adapter.connect(userA).withdraw(amount, userA.address, userA.address)
+
+              const currentBalance = await token.balanceOf(userA.address)
+              expect(currentBalance).to.be.equal(previousBalance.add(amount))
+
+              await expect(erc4626Adapter.connect(userB).withdraw(amount, userB.address, userB.address)).not.to.be
+                .reverted
             })
 
-            totalAssets = totalAssets.add(amount) // 660
-            shareValue = fp(5.32)
-            userAAssets = calculateUserAssets(userAShares) // 532
-            userBAssets = calculateUserAssets(userBShares) // 57
-            userCShares = userCShares.add(fp(33).mul(fp(1)).div(shareValue)) // 13.34
-            userCAssets = calculateUserAssets(userCShares) // 71
-            ;(totalShares = userAShares.add(userBShares).add(userCShares)), // 124.05
-              checkStatus({
-                totalAssets,
-                shareValue,
-                userAShares,
-                userAAssets,
-                userBShares,
-                userBAssets,
-                userCShares,
-                userCAssets,
-                totalShares,
-                previousTotalAssets,
-              })
+            it(`${fee ? 'charges a fee' : 'does not charge any fee'}`, async function () {
```

Comment:
Let's always use arrow functions unless we do need `this` bindings:

```suggestion
            it(`${fee ? 'charges a fee' : 'does not charge any fee'}`, async () => {
```